### PR TITLE
add support for sync_bn

### LINKED ISSUE #2589: Add synced batchnorm support
https://pytorch.org/docs/stable/nn.html#torch.nn.SyncBatchNorm

image

@justusschock or @awaelchli interested?

### PR DESCRIPTION
**Pull request by ananyahjha93**
## What does this PR do?

Adds support for global batch norm using sync_bn and allows for customizing sync_bn with a provision to override configure_sync_bn() function in LightningModule.

Fixes #2589, #2509 

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -957,6 +957,30 @@ def init_ddp_connection(self, global_rank: int, world_size: int, is_slurm_managi
...
+
+        Args:
+            model: pointer to current :class:`LightningModule`.
+            backend: select between 'torch' and 'apex'.
+
+        Return:
+            LightningModule with batchnorm layers converted to sync batchnorm
+        """
+        # process_group = None defaults to whole world in both cases
+        if backend == 'torch':
```
Comment (justusschock) [inline pytorch_lightning/core/lightning.py:975]: What's the difference of these two backends? I.e. if they're the same, should we really add another dependency on apex? Thought that we try to remove it from dependencies, now that pt has also native amp support (of course being backwards compatible with amp, but not introduce a new dependency with a new feature)

Comment (ananyahjha93) [reply pytorch_lightning/core/lightning.py:976]: There were a few issues with pytorch's implementation and hence the authors of swav were using apex for this. Also, they found apex to be faster.

File: pl_examples/basic_examples/sync_bn.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+import pytorch_lightning as pl
+
+import torchvision.transforms as transforms
+from torchvision.datasets import MNIST
+from torch.utils.data import random_split
+from torch.utils.data import DataLoader
+
+
+pl.seed_everything(234)
+EPSILON = 1e-12
```
Comment (Borda) [inline pl_examples/basic_examples/sync_bn.py:24]: consider using https://github.com/PyTorchLightning/pytorch-lightning/blob/a55c481d5d27028bd34024705d79cbea66c48d32/pytorch_lightning/utilities/__init__.py#L19

File: pl_examples/basic_examples/sync_bn.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        ])
+
+        # self.dims is returned when you call dm.size()
+        # Setting default dims here because we know them.
+        # Could optionally be assigned dynamically in dm.setup()
+        self.dims = (1, 28, 28)
+
+    def prepare_data(self):
+        # download only
+        MNIST(os.getcwd(), train=True, download=True, transform=transforms.ToTensor())
```
Comment (Borda) [inline pl_examples/basic_examples/sync_bn.py:45]: let's use datamodule instead 
cc @nateraw

File: pytorch_lightning/core/lightning.py [hunk 4]
```diff
@@ -957,6 +957,38 @@ def init_ddp_connection(self, global_rank: int, world_size: int, is_slurm_managi
...
+
+        # process_group = None defaults to whole world in both cases
+        if backend == 'torch':
+            model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model, process_group=None)
+        elif backend == 'apex':
+            import apex
+
+            model = apex.parallel.convert_syncbn_model(model, process_group=None)
+        else:
+            raise ValueError("only torch and apex options are supported for sync_batchnorm at the moment.")
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:988]:
```suggestion
            raise ValueError("only `torch` and `apex` options are supported for `sync_batchnorm` at the moment.")
```

File: pl_examples/basic_examples/sync_bn.py (hunk 2 above)
Comment (ananyahjha93) [reply pl_examples/basic_examples/sync_bn.py:24]: @Borda this comes out to be quite a large value '0.000977'. numpy.finfo(numpy.float32).eps is 1.1920929e-07.

Comment (Borda) [reply pl_examples/basic_examples/sync_bn.py:24]: yes, but 1e-12 fro AMP with float16 is 0
cc: @williamFalcon

**Comment by ananyahjha93** [comment]
@Borda @williamFalcon removed apex option as backend, the tests for apex were initially passing because its sync_bn was falling back on pytorch's default version. When I reinstalled apex and got it to call its own sync_bn, there were quite a few issues tensors syncing between GPUs. 

Basic idea, the ```configure_sync_bn``` function provides a default torch implementation and can be overridden of specific versions are required.

**Comment by ananyahjha93** [comment]
The DDP script tests were not working as of now, so I have an example in pl_examples/basic_examples which verifies the correct working of sync batch-norm.

**Comment by williamFalcon** [comment]
sorry, why not put as a test using ddp spawn?

**Comment by ananyahjha93** [comment]
@nateraw ```test_full_loop_ddp_spawn``` is failing in drone because the test accuracy is not greater than 0.8. tests/core/test_datamodules.py
